cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -791,9 +884,54 @@ public ApplicationWithPrograms deployApp(NamespaceId namespace, @Nullable String
     if (artifactDetail.isEmpty()) {
       throw new ArtifactNotFoundException(range.getNamespace(), range.getName());
     }
-    return deployApp(namespace, appName, appVersion, configStr, programTerminator, artifactDetail.iterator().next(),
-                     ownerPrincipal, updateSchedules == null ? appUpdateSchedules : updateSchedules, isPreview,
-                     userProps);
+    return deployApp(namespace, appName, appVersion, configStr, changeSummary, programTerminator,
+                     artifactDetail.iterator().next(), ownerPrincipal, updateSchedules == null ?
+                     appUpdateSchedules : updateSchedules, isPreview, userProps);
+  }
+
+  /**
+   * Deploy an application using the specified artifact and configuration. When an app is deployed, the Application
+   * class is instantiated and configure() is called in order to generate an {@link ApplicationSpecification}.
+   * Programs, datasets, and streams are created based on the specification before the spec is persisted in the
+   * {@link Store}. This method can create a new application as well as update an existing one.
+   *
+   * @param namespace the namespace to deploy the app to
+   * @param appName the name of the app. If null, the name will be set based on the application spec
+   * @param summary the artifact summary of the app
+   * @param configStr the configuration to send to the application when generating the application specification
+   * @param programTerminator a program terminator that will stop programs that are removed when updating an app.
+   *                          For example, if an update removes a flow, the terminator defines how to stop that flow.
+   * @param ownerPrincipal the kerberos principal of the application owner
+   * @param updateSchedules specifies if schedules of the workflow have to be updated,
+   *                        if null value specified by the property "app.deploy.update.schedules" will be used.
+   * @param isPreview whether the app deployment is for preview
+   * @param userProps the user properties for the app deployment, this is basically used for preview deployment
+   * @return information about the deployed application
+   * @throws InvalidArtifactException if the artifact does not contain any application classes
+   * @throws IOException if there was an IO error reading artifact detail from the meta store
+   * @throws ArtifactNotFoundException if the specified artifact does not exist
+   * @throws Exception if there was an exception during the deployment pipeline. This exception will often wrap
+   *                   the actual exception
+   */
+  public ApplicationWithPrograms deployApp(NamespaceId namespace, @Nullable String appName, @Nullable String appVersion,
```

Comment:
we shouldn't have so many deployApp() methods. Do they only differ in which optional arguments they take?

If so, we should refactor things so that the full set of arguments is in some separate class like an AppDeploymentRequest class (or could we use AppDeploymentInfo here?), and just have a single deployApp() method here.

Reply:
yes, they differ in the optional parameters - I can refactor them, however this will again require modifying a bunch of tests - just for the optional parameters